lib/mariaex/row_parser.ex

```diff
@@ -0,0 +1,216 @@
+defmodule Mariaex.RowParser do
+  @moduledoc """
+  Parse a row of the MySQL protocol
+
+  This parser makes extensive use of binary pattern matching and recursion to take advantage
+  of Erlang's optimizer that will not create sub binaries when called recusively.
+  """
+  use Bitwise
+
+  def decode_bin_rows(packet, fields) do
+    nullbin_size = div(length(fields) + 7 + 2, 8)
+    decode_bin_rows(packet, fields, nullbin_size)
+  end
+
+  defp decode_bin_rows(packet, fields, nullbin_size) do
+    << 0 :: 8, null_bitfield :: size(nullbin_size)-little-unit(8), rest :: binary >> = packet
+    decode_bin_rows(rest, fields, null_bitfield >>> 2, [])
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [_ | fields], nullint, acc) when (nullint &&& 1) === 1 do
+    decode_bin_rows(rest, fields, nullint >>> 1, [nil | acc])
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:string | fields], null_bitfield,  acc) do
+    decode_string(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:uint8 | fields], null_bitfield, acc) do
+    decode_uint8(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:int8 | fields], null_bitfield, acc) do
+    decode_int8(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:uint16 | fields], null_bitfield, acc) do
+    decode_uint16(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:int16 | fields], null_bitfield, acc) do
+    decode_int16(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:uint32 | fields], null_bitfield, acc) do
+    decode_uint32(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:int32 | fields], null_bitfield, acc) do
+    decode_int32(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:uint64 | fields], null_bitfield, acc) do
+    decode_uint64(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:int64 | fields], null_bitfield, acc) do
+    decode_int64(rest, fields, null_bitfield >>> 1, acc)
+  end
+
+  defp decode_bin_rows(<<rest::bits>>, [:year | fields], null_bitfield, acc) do
+    decode_int16(rest, fields, null_bitfield >>> 1, acc)
```

Comment:
Of course the signed bit never comes in given the range of mysql date range but technically this is unsigned not signed. Could use `:uint16` as field instead to avoid the extra clause?